### PR TITLE
binutils: 2.40 -> 2.41

### PR DESCRIPTION
**Pull request by trofi**
Changes: https://lists.gnu.org/archive/html/info-gnu/2023-07/msg00009.html

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by trofi** [comment]
It's a draft PR as it has a few known problems:

- [ ] `ffmpeg` needs an assembly tweak: https://github.com/NixOS/nixpkgs/pull/246630
- [ ] `pkgsCross.mips64-linux-gnuabi64.stdenv` fails to apply patches:

      applying patch /nix/store/ydc99rr8wayfylb61l3jgx1yqh7n6236-mips64-default-n64.patch
      patching file bfd/config.bfd
      Hunk #1 FAILED at 927.

**Comment by Atemu** [comment]
This successfully builds up to ffmpeg-headless with https://github.com/NixOS/nixpkgs/pull/246630 on aarch64-darwin.

**Review by Atemu — COMMENTED** [review]
This successfully builds up to ffmpeg-headless with https://github.com/NixOS/nixpkgs/pull/246630 on aarch64-darwin.

**Comment by Atemu** [comment]
Um, did you intend on closing this one aswell?

**Comment by trofi** [comment]
Yeah, I don't plan to fix all of `nixpkgs` for `binutils-2.41` support any time soon.

**Comment by Atemu** [comment]
Understandable but I don't think anyone expects you to do that.

If some critical core package breaks because of this, sure we might want to take a closer look/post-pone the binutils upgrade but for leaf-y packages, I'd say it's up to the maintainers of those individual packages to fix them for binutils 2.41. The most you could be expected to do there IMO is notify the maintainers and mark the package as broken if they don't respond but even that isn't commonly done (if at all).  
You already went out of your way to fix ffmpeg (I'd classify that as a critical core package), that's already more than I could expect you to do IMO.

I know you build tons of stuff all the time, so you probably have the resources to easily do this: If you can build a basic NixOS closure on x86 with this change, I think that'd be good enough for a start. Perhaps we'd want to get a full hydra eval on this going to make sure we haven't broken some niche in a major way but I don't know how much potential there is for breakage from binutils, so that might not even be necessary.

**Comment by trofi** [comment]
Regarding this specific PR:

I did build my `system` attribute with it and `ffmpeg` was the only failure. Among cross-compilers `binutils` is broken at least on `mips64` due to a `nixpkgs` downstream patch I'd rather not touch for non-technical reasons. It's clearly not ready for merge and I don't plan to fix it. But it's good enough for my local use.

Regarding large scale changes in general:

My small experience of large-scale changes in `nixpkgs` tells me that if there is a chance of breaking ~100+ packages it's better to provide most of fixes upfront. As otherwise retroactive fixes are burdensome on `staging-next` maintainers if it's more than 2-3 broken packages.

Getting upfront fixes usually require a hydra jobset and follow-up fixes. For 100+ breakages maintainer notifications don't work and sometimes are actively discouraged (which is understandable: being CCed in a PR with 100+ changes when you care about only one is not very productive). And reality is that `nixpkgs` maintainers are not really expected to actively fix their packages until ZHF (twice a year). Fixes are almost always done by someone else.

On top of that my preference for large scale changes is to keep `nixpkgs` in bisectable state (as in, `staging` branch should not introduce known breakages).

In the past I used to provide small incremental fixes into `master` and `staging` for individual packages before a final disrupting change is landed. For example in case of `ncurses` update it was 200-300 changes over the course of a few months. While it works perfectly for me it does not always work for reviewers as people understandably want to see the failure before they accept the fix.

Thus nowadays I opt for providing fixes to upstreams directly and let them trickle naturally into `nixpkgs`. It takes way longer to get there (and not all upstreams are alive). But at least if reduces some load on `nixpkgs` reviewers.

Tl;DR of the above is: no, there is an expectation to do quite a bit of work for large-scale changes in `nixpkgs` one way or another.

**Comment by trofi** [comment]
Filed https://github.com/NixOS/nixpkgs/issues/262159 for someone to update `binutils` to `2.41`.

**Comment by trofi** [comment]
I'll try again: https://github.com/NixOS/nixpkgs/pull/283732
